File: frontend/utils/physicsEngine.ts
// Physics engine wrapper using Cannon.js for realistic spaceship movement

import * as CANNON from "cannon-es";

export interface PhysicsState {
  position: { x: number; y: number; z: number };
  velocity: { x: number; y: number; z: number };
  rotation: { x: number; y: number; z: number };
  angularVelocity: { x: number; y: number; z: number };
}

export interface SpaceshipPhysicsConfig {
  mass: number;
  maxThrust: number;
  maxAngularVelocity: number;
  damping: number;
  angularDamping: number;
  dragCoefficient: number;
}

export const defaultSpaceshipConfig: SpaceshipPhysicsConfig = {
  mass: 1000, // kg
  maxThrust: 5000, // N
  maxAngularVelocity: 2.0, // rad/s
  damping: 0.8, // Linear damping
  angularDamping: 0.9, // Angular damping
  dragCoefficient: 0.1, // Air resistance
};

export class SpaceshipPhysicsEngine {
  private world: CANNON.World;
  private spaceshipBody: CANNON.Body;
  private config: SpaceshipPhysicsConfig;
  private targetPosition: CANNON.Vec3;
  private targetVelocity: CANNON.Vec3;
  private lastUpdateTime: number;

  constructor(config: SpaceshipPhysicsConfig = defaultSpaceshipConfig) {
    this.config = config;
    
    // Create physics world
    this.world = new CANNON.World();
    this.world.gravity.set(0, 0, 0); // Zero gravity for space
    this.world.broadphase = new CANNON.NaiveBroadphase();
    this.world.solver.iterations = 10;

    // Create spaceship body (box shape for simplicity)
    const shape = new CANNON.Box(new CANNON.Vec3(1, 0.5, 0.3)); // Approximate spaceship dimensions
    this.spaceshipBody = new CANNON.Body({ mass: config.mass });
    this.spaceshipBody.addShape(shape);
    this.spaceshipBody.linearDamping = config.damping;
    this.spaceshipBody.angularDamping = config.angularDamping;
    this.spaceshipBody.position.set(0, 0, 0);
    
    this.world.addBody(this.spaceshipBody);

    // Initialize target tracking
    this.targetPosition = new CANNON.Vec3(0, 0, 0);
    this.targetVelocity = new CANNON.Vec3(0, 0, 0);
    this.lastUpdateTime = performance.now();
  }

  /**
   * Set target position for navigation
   */
  setTarget(x: number, y: number, z: number = 0): void {
    this.targetPosition.set(x, y, z);
  }

  /**
   * Apply thrust force based on agent action
   */
  applyThrust(action: "IDLE" | "GLIDE" | "BOOST" | "STABILIZE", intensity: number = 1.0): void {
    const force = new CANNON.Vec3();
    
    // Calculate direction to target
    const direction = new CANNON.Vec3();
    direction.sub(this.targetPosition, this.spaceshipBody.position);
    const distance = direction.length();
    
    if (distance > 0.1) {
      direction.normalize();
      
      switch (action) {
        case "BOOST":
          // Maximum thrust towards target
          const boostForce = this.config.maxThrust * intensity * 1.5;
          force.set(
            direction.x * boostForce,
            direction.y * boostForce,
            direction.z * boostForce
          );
          break;
          
        case "GLIDE":
          // Minimal thrust, mostly coasting
          const glideForce = this.config.maxThrust * intensity * 0.3;
          force.set(
            direction.x * glideForce,
            direction.y * glideForce,
            direction.z * glideForce
          );
          // Apply drag
          const velocity = this.spaceshipBody.velocity;
          const dragForce = velocity.scale(-this.config.dragCoefficient * velocity.length());
          force.vadd(dragForce, force);
          break;
          
        case "STABILIZE":
          // Counteract angular velocity and apply stabilizing thrust
          const stabilizeForce = this.config.maxThrust * intensity * 0.5;
          force.set(
            direction.x * stabilizeForce,
            direction.y * stabilizeForce,
            direction.z * stabilizeForce
          );
          
          // Apply angular damping for stabilization
          this.spaceshipBody.angularVelocity.scale(0.7, this.spaceshipBody.angularVelocity);
          break;
          
        case "IDLE":
        default:
          // Just apply drag
          const idleVelocity = this.spaceshipBody.velocity;
          const idleDrag = idleVelocity.scale(-this.config.dragCoefficient * idleVelocity.length());
          force.copy(idleDrag);
          break;
      }
      
      this.spaceshipBody.applyForce(force, this.spaceshipBody.position);
    }
  }

  /**
   * Apply torque for banking/rolling based on movement direction
   */
  applyBanking(desiredRoll: number, rollFactor: number = 0.1): void {
    const currentRoll = this.getRoll();
    const rollDifference = desiredRoll - currentRoll;
    
    // Apply torque to achieve desired roll
    const torque = new CANNON.Vec3(0, 0, rollDifference * rollFactor * 100);
    this.spaceshipBody.torque.copy(torque);
    
    // Limit angular velocity
    if (this.spaceshipBody.angularVelocity.length() > this.config.maxAngularVelocity) {
      this.spaceshipBody.angularVelocity.normalize();
      this.spaceshipBody.angularVelocity.scale(this.config.maxAngularVelocity, this.spaceshipBody.angularVelocity);
    }
  }

  /**
   * Step the physics simulation
   */
  step(deltaTime: number = 0.016): void {
    // Clamp deltaTime to prevent large jumps
    const clampedDelta = Math.min(deltaTime, 0.1);
    this.world.step(clampedDelta);
  }

  /**
   * Get current physics state
   */
  getState(): PhysicsState {
    return {
      position: {
        x: this.spaceshipBody.position.x,
        y: this.spaceshipBody.position.y,
        z: this.spaceshipBody.position.z,
      },
      velocity: {
        x: this.spaceshipBody.velocity.x,
        y: this.spaceshipBody.velocity.y,
        z: this.spaceshipBody.velocity.z,
      },
      rotation: {
        x: this.spaceshipBody.quaternion.x,
        y: this.spaceshipBody.quaternion.y,
        z: this.spaceshipBody.quaternion.z,
      },
      angularVelocity: {
        x: this.spaceshipBody.angularVelocity.x,
        y: this.spaceshipBody.angularVelocity.y,
        z: this.spaceshipBody.angularVelocity.z,
      },
    };
  }

  /**
   * Get roll angle in radians
   */
  getRoll(): number {
    // Extract roll from quaternion
    const q = this.spaceshipBody.quaternion;
    const sinr_cosp = 2 * (q.w * q.x + q.y * q.z);
    const cosr_cosp = 1 - 2 * (q.x * q.x + q.y * q.y);
    return Math.atan2(sinr_cosp, cosr_cosp);
  }

  /**
   * Get pitch angle in radians
   */
  getPitch(): number {
    const q = this.spaceshipBody.quaternion;
    const sinp = 2 * (q.w * q.y - q.z * q.x);
    if (Math.abs(sinp) >= 1) {
      return Math.sign(sinp) * Math.PI / 2;
    }
    return Math.asin(sinp);
  }

  /**
   * Get yaw angle in radians
   */
  getYaw(): number {
    const q = this.spaceshipBody.quaternion;
    const siny_cosp = 2 * (q.w * q.z + q.x * q.y);
    const cosy_cosp = 1 - 2 * (q.y * q.y + q.z * q.z);
    return Math.atan2(siny_cosp, cosy_cosp);
  }

  /**
   * Get speed in m/s
   */
  getSpeed(): number {
    return this.spaceshipBody.velocity.length();
  }

  /**
   * Reset physics state
   */
  reset(x: number = 0, y: number = 0, z: number = 0): void {
    this.spaceshipBody.position.set(x, y, z);
    this.spaceshipBody.velocity.set(0, 0, 0);
    this.spaceshipBody.angularVelocity.set(0, 0, 0);
    this.spaceshipBody.quaternion.set(0, 0, 0, 1);
  }

  /**
   * Cleanup
   */
  dispose(): void {
    this.world.removeBody(this.spaceshipBody);
    // Note: Cannon.js doesn't have explicit dispose, but we can clear references
  }
}

File: native/physics/src/physics_engine.cpp
#include "physics_engine.hpp"
#include <algorithm>
#include <cmath>

namespace aeronav {

PhysicsEngine::PhysicsEngine()
    : body_()
    , config_()
    , targetPosition_(Vector3::zero())
{
    // Apply default config to body
    body_.setMass(config_.mass);
    body_.setLinearDamping(config_.linearDamping);
    body_.setAngularDamping(config_.angularDamping);
    body_.setDragCoefficient(config_.dragCoefficient);
    body_.setMaxAngularVelocity(config_.maxAngularVelocity);
}

PhysicsEngine::PhysicsEngine(const SpaceshipConfig& config)
    : body_()
    , config_(config)
    , targetPosition_(Vector3::zero())
{
    // Apply config to body
    body_.setMass(config_.mass);
    body_.setLinearDamping(config_.linearDamping);
    body_.setAngularDamping(config_.angularDamping);
    body_.setDragCoefficient(config_.dragCoefficient);
    body_.setMaxAngularVelocity(config_.maxAngularVelocity);
}

void PhysicsEngine::step(float deltaTime) {
    // Clamp delta time to prevent instability (matching JS: max 0.1s)
    deltaTime = std::min(deltaTime, 0.1f);

    if (deltaTime <= 0.0f) return;

    // Integrate physics
    body_.integrate(deltaTime);
}

void PhysicsEngine::reset(float x, float y, float z) {
    body_.reset();
    body_.setPosition(Vector3(x, y, z));
    targetPosition_ = Vector3::zero();
}

void PhysicsEngine::setTarget(float x, float y, float z) {
    targetPosition_ = Vector3(x, y, z);
}

Vector3 PhysicsEngine::getDirectionToTarget() const {
    Vector3 direction = targetPosition_ - body_.getPosition();
    float length = direction.length();
    if (length > 1e-6f) {
        return direction / length;
    }
    return Vector3::zero();
}

float PhysicsEngine::getDistanceToTarget() const {
    return (targetPosition_ - body_.getPosition()).length();
}

void PhysicsEngine::applyDragForce() {
    Vector3 velocity = body_.getVelocity();
    float speed = velocity.length();
    if (speed > 1e-6f) {
        // Drag force proportional to velocity squared (magnitude) in opposite direction
        float dragMagnitude = config_.dragCoefficient * speed;
        Vector3 dragForce = velocity.normalized() * (-dragMagnitude);
        body_.applyForce(dragForce);
    }
}

void PhysicsEngine::applyThrust(ThrustAction action, float intensity) {
    float distance = getDistanceToTarget();

    // Don't apply thrust if very close to target (matching JS: 0.1)
    if (distance <= 0.1f) return;

    Vector3 direction = getDirectionToTarget();
    Vector3 force = Vector3::zero();

    switch (action) {
        case ThrustAction::BOOST:
            // Maximum thrust towards target (1.5x multiplier)
            {
                float boostForce = config_.maxThrust * intensity * 1.5f;
                force = direction * boostForce;
            }
            break;

        case ThrustAction::GLIDE:
            // Minimal thrust, mostly coasting (0.3x multiplier)
            {
                float glideForce = config_.maxThrust * intensity * 0.3f;
                force = direction * glideForce;
                // Apply additional drag during glide
                applyDragForce();
            }
            break;

        case ThrustAction::STABILIZE:
            // Stabilizing thrust with angular damping (0.5x multiplier)
            {
                float stabilizeForce = config_.maxThrust * intensity * 0.5f;
                force = direction * stabilizeForce;

                // Apply extra angular damping for stabilization
                Vector3 angVel = body_.getAngularVelocity();
                body_.setAngularVelocity(angVel * 0.7f);
            }
            break;

        case ThrustAction::IDLE:
        default:
            // Just apply drag
            applyDragForce();
            break;
    }

    if (force.lengthSquared() > 1e-6f) {
        body_.applyForce(force);
    }
}

void PhysicsEngine::applyThrustByName(int action, float intensity) {
    ThrustAction thrustAction;
    switch (action) {
        case 0: thrustAction = ThrustAction::IDLE; break;
        case 1: thrustAction = ThrustAction::GLIDE; break;
        case 2: thrustAction = ThrustAction::BOOST; break;
        case 3: thrustAction = ThrustAction::STABILIZE; break;
        default: thrustAction = ThrustAction::IDLE; break;
    }
    applyThrust(thrustAction, intensity);
}

void PhysicsEngine::applyBanking(float desiredRoll, float rollFactor) {
    float currentRoll = getRoll();
    float rollDifference = desiredRoll - currentRoll;

    // Apply torque to achieve desired roll (matching JS implementation)
    Vector3 torque(0.0f, 0.0f, rollDifference * rollFactor * 100.0f);
    body_.applyTorque(torque);
}

PhysicsState PhysicsEngine::getState() const {
    PhysicsState state;
    state.position = body_.getPosition();
    state.velocity = body_.getVelocity();

    // Convert quaternion to Euler for compatibility with JS frontend
    // Note: JS stores quaternion x,y,z in rotation (not w), so we match that
    Quaternion q = body_.getRotation();
    state.rotation = Vector3(q.x, q.y, q.z);

    state.angularVelocity = body_.getAngularVelocity();
    return state;
}

float PhysicsEngine::getRoll() const {
    Quaternion q = body_.getRotation();
    float sinr_cosp = 2.0f * (q.w * q.x + q.y * q.z);
    float cosr_cosp = 1.0f - 2.0f * (q.x * q.x + q.y * q.y);
    return std::atan2(sinr_cosp, cosr_cosp);
}

float PhysicsEngine::getPitch() const {
    Quaternion q = body_.getRotation();
    float sinp = 2.0f * (q.w * q.y - q.z * q.x);
    if (std::abs(sinp) >= 1.0f) {
        return std::copysign(3.14159265358979f * 0.5f, sinp);
    }
    return std::asin(sinp);
}

float PhysicsEngine::getYaw() const {
    Quaternion q = body_.getRotation();
    float siny_cosp = 2.0f * (q.w * q.z + q.x * q.y);
    float cosy_cosp = 1.0f - 2.0f * (q.y * q.y + q.z * q.z);
    return std::atan2(siny_cosp, cosy_cosp);
}

float PhysicsEngine::getSpeed() const {
    return body_.getVelocity().length();
}

void PhysicsEngine::setConfig(const SpaceshipConfig& config) {
    config_ = config;
    body_.setMass(config_.mass);
    body_.setLinearDamping(config_.linearDamping);
    body_.setAngularDamping(config_.angularDamping);
    body_.setDragCoefficient(config_.dragCoefficient);
    body_.setMaxAngularVelocity(config_.maxAngularVelocity);
}

void PhysicsEngine::setMass(float mass) {
    config_.mass = mass;
    body_.setMass(mass);
}

void PhysicsEngine::setMaxThrust(float thrust) {
    config_.maxThrust = thrust;
}

void PhysicsEngine::setMaxAngularVelocity(float maxAngVel) {
    config_.maxAngularVelocity = maxAngVel;
    body_.setMaxAngularVelocity(maxAngVel);
}

void PhysicsEngine::setLinearDamping(float damping) {
    config_.linearDamping = damping;
    body_.setLinearDamping(damping);
}

void PhysicsEngine::setAngularDamping(float damping) {
    config_.angularDamping = damping;
    body_.setAngularDamping(damping);
}

void PhysicsEngine::setDragCoefficient(float drag) {
    config_.dragCoefficient = drag;
    body_.setDragCoefficient(drag);
}

} // namespace aeronav
